### PR TITLE
raftstore: add region_id to snap concurrency limiter APIs

### PR DESCRIPTION
**Pull request by hbisheng**
### What is changed and how it works?

Issue Number: Ref #15972

What's Changed:

```commit-message
An issue with the previous implementation of the concurrency limiter is that 
its APIs are not idempotent. If a single region leader keeps sending precheck 
requests to the same receiver (this may happen if the precheck response is 
somehow lost on the network), it would consume all reservations, thereby 
blocking all other snapshots. This commit addresses this problem by updating 
the concurrency limiter to deduplicate requests based on `region_id`. If the 
limiter already has a valid reservation for a region, calling try_recv() again 
will not allocate a new one. 
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by hbisheng** [comment]
I'd appreciate a review of this /cc @Connor1996 @overvenus

File: components/raftstore/src/store/snap.rs [hunk 1]
```diff
@@ -2009,46 +2008,53 @@ impl SnapRecvConcurrencyLimiter {
...
-        {
-            timestamps.pop_front();
-        }
+    fn evict_expired_timestamps(
+        &self,
+        timestamps: &mut HashMap<u64, Instant>,
+        current_time: Instant,
+    ) {
+        timestamps.retain(|_, timestamp| {
+            current_time.duration_since(timestamp.clone()) <= Duration::from_secs(self.ttl_secs)
```
Comment (overvenus) [inline components/raftstore/src/store/snap.rs:2047]: Maybe we should add a log for this unusual case?

**Review by overvenus — APPROVED** [review]

File: components/raftstore/src/store/snap.rs (hunk 1 above)
Comment (hbisheng) [reply components/raftstore/src/store/snap.rs]: added a warning for it

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**